core/task_manager.go

```diff
@@ -32,11 +35,134 @@ import (
+}
+
+// This function updates the state of a task, setting its status as necessary.
+// By modern standards, this function may seem long, but I don't think breaking
+// it into disparate pieces makes it any easier to understand, given the state
+// information being managed--that just creates other abstraction problems.
+func (task *taskType) Update() error {
+	username := "user" // FIXME: how do we obtain this from our Orcid ID?
+	sourceEndpoint, err := task.Source.Endpoint()
+	if err != nil {
+		return err
+	}
+	destinationEndpoint, err := task.Destination.Endpoint()
+	if err != nil {
+		return err
+	}
+
+	if task.Resources == nil { // new task!
+		// resolve file paths using file IDs
+		task.Resources, err = task.Source.Resources(task.FileIds)
+		if err == nil {
+			// tell the source DB to stage the files, stash the task, and return
+			// its new ID
+			task.Staging.UUID, err = task.Source.StageFiles(task.FileIds)
+			task.Staging.Valid = true
+			if err == nil {
+				task.Status = TransferStatus{
+					Code:     TransferStatusStaging,
+					NumFiles: len(task.FileIds),
+				}
+			}
+		}
+	} else if task.Staging.Valid { // we're staging
+		// are the files staged?
+		var staged bool
+		staged, err = sourceEndpoint.FilesStaged(task.Resources)
+		if err == nil && staged {
+			// initiate the transfer
+			fileXfers := make([]FileTransfer, len(task.Resources))
+			for i, resource := range task.Resources {
+				destinationPath := filepath.Join(username, task.Id.String(), resource.Path)
+				fileXfers[i] = FileTransfer{
+					SourcePath:      resource.Path,
+					DestinationPath: destinationPath,
+					Hash:            resource.Hash,
+				}
+			}
+			task.Transfer.UUID, err = sourceEndpoint.Transfer(destinationEndpoint, fileXfers)
+			if err == nil {
+				task.Status = TransferStatus{
+					Code:     TransferStatusActive,
+					NumFiles: len(task.FileIds),
+				}
+				task.Staging = uuid.NullUUID{}
+				task.Transfer.Valid = true
+			}
+		}
+	} else if task.Transfer.Valid { // we're transferring
+		// has the data transfer completed?
+		var xferStatus TransferStatus
+		xferStatus, err = sourceEndpoint.Status(task.Transfer.UUID)
+		if err == nil && (xferStatus.Code == TransferStatusSucceeded ||
+			xferStatus.Code == TransferStatusFailed) { // transfer finished
+			task.Transfer = uuid.NullUUID{}
+			if xferStatus.Code == TransferStatusSucceeded {
+				// generate a manifest for the transfer
+				manifest := DataPackage{
+					Name:      "manifest",
+					Resources: make([]DataResource, len(task.Resources)),
+				}
+				copy(manifest.Resources, task.Resources)
+
+				// write the manifest to disk and begin transferring it to the
+				// destination endpoint
+				var manifestBytes []byte
+				manifestBytes, err = json.Marshal(manifest)
+				if err == nil {
+					var manifestFile *os.File
+					manifestFile, err = os.CreateTemp(task.LocalEndpoint.Root(), "manifest.json")
+					if err == nil {
```

Comment:
all the nesting starts making the code a lot less readable -- could you do something like
```go
if err != nil {
  return err
}
```

to keep it in check, instead of the `if` cascade?

Reply:
or is there something like a `while` construct that you could use - `while err == nil ...`?

Reply:
I don't think `while` can be used that way, but I can certainly use your first suggestion.

Reply:
Do you mind if I change this error checking style in a subsequent PR? I think there's enough of this in the code that it might be easiest to treat as its own thing.

Reply:
yes, no problem

Reply:
> or is there something like a `while` construct that you could use - `while err == nil ...`?
> 
> This might just be a go idiom I have to get used to, but it makes the program flow look very disjointed and repetitive when you're always checking for errors (one positive about having systems where errors are thrown, stopping the flow of execution).

Yes, there remains a huge debate between the error code folk and the exception folk, and in this respect Go falls on a different side from Java and Python.

Reply:
See issue #38